lib/java/persistance/src/main/java/org/enso/persist/PerInputImpl.java

```diff
@@ -241,17 +254,55 @@ static <T> Reference<T> readIndirectAsReference(
     }
   }
 
-  static final record InputCache(
-      ByteBuffer buf,
-      Function<Object, Object> readResolve,
-      Map<Integer, Object> cache,
-      PerMap map) {
-    InputCache(ByteBuffer buf, Function<Object, Object> readResolve) {
-      this(
-          buf,
-          readResolve == null ? Function.identity() : readResolve,
-          new HashMap<>(),
-          PerMap.create());
+  static final class InputCache {
+    private final Map<Integer, Object> cache = new HashMap<>();
+    private final Function<Object, Object> readResolve;
+    private final PerMap map;
+    private final ByteBuffer buf;
+    private final Reference[] refs;
+
+    private InputCache(
+        ByteBuffer buf, Function<Object, Object> readResolve, PerMap map, int[] refs) {
+      this.buf = buf;
+      this.readResolve = readResolve;
+      this.map = map;
+      this.refs = new Reference[refs.length];
+      for (var i = 0; i < refs.length; i++) {
+        if (refs[i] < 0) {
+          this.refs[i] = this.refs[0];
+        } else {
+          this.refs[i] = PerBufferReference.cached(null, this, refs[i]);
+        }
+      }
```

Comment:
Can we get some short documentation summary describing the idea behind the `PerBufferReference.cached` and this `InputCache`?

I don't quite see what is the reasoning here, what does it mean for `refs[i] < 0` and why then it's replaced with `refs[0]` etc.

I guess if I took 1-2 hours to try to understand this code I could reverse engineer the idea what is happening here, but I think there should be at least minimal documentation describing the intent here, so that anyone reading the code can have at least a general idea of what is happening here.

Reply:
The idea is clearly visible in this commit 03494b8979d536f6d7493f8565c81269f1e45077 which is associated with following comment https://github.com/enso-org/enso/pull/9972#discussion_r1604813584

> took 1-2 hours to try to understand this code

No need to spend so much time. Enough to use `git blame` and see [it](https://github.com/enso-org/enso/pull/9972/commits/03494b8979d536f6d7493f8565c81269f1e45077) immediately.